### PR TITLE
Revert Envoy xDS updates on regeneration failure

### PR DESCRIPTION
**Pull request by rlenglet**
Extend the xDS interfaces to return a revert function for every xDS cache update.
In `regenerateBPF`, maintain a stack of "revert" functions to revert the side-effects of the regeneration on failure. Also maintain a list of "finalize" functions that are executed only if regeneration completes successfully, to complete actions that are not reversible.
Revert xDS network policy changes and proxy redirect addition/removal on regeneration failure.

The effects of this PR is that all the side-effects on L7 proxies of an endpoint's BPF regeneration are reverted on regeneration failure, including all Envoy-related state, esp. the xDS cache updates.
No other regeneration side-effects are reverted. Notably, changes to the BPF policy map are not reverted, which can result in traffic loss if the policy map ends up not being consistent with the configuration of the L7 proxies after reverting.

Fixes: #5317
Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5451)

### CODE REVIEW

**Comment by tgraf** [comment]
How and when does this help? The PUT endpoint API has a timeout as well while building synchronously so we will just see more aborts from there followed by endpoint deletions instead.

**Comment by rlenglet** [comment]
> we will just see more aborts from there followed by endpoint deletions instead.

Why will we see more aborts? BPF compilation already ignores context cancelations.

File: pkg/envoy/xds/cache.go
Comment (jrajahalme) [inline pkg/envoy/xds/cache.go:121]: I'd call this variable `revertChangedResources`...

File: pkg/envoy/xds/cache.go
Comment (jrajahalme) [inline pkg/envoy/xds/cache.go:128]: And this one `deleteInsertedNames`?

**Review by jrajahalme — APPROVED** [review]
LGTM

**Comment by tgraf** [comment]
> Why will we see more aborts? BPF compilation already ignores context cancelations.

`func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder {` will just delete the endpoint after a`endpoint.EndpointGenerationTimeout` timeout. Regardless of whether the regeneration is still ongoing or not.

**Comment by rlenglet** [comment]
Is your concern that we'd wait for up to 55 seconds after BPF compilation? I agree that could be too long. How about reducing that Envoy ACK timeout, like we were doing already in `deleteEndpointQuiet` (it was 10 seconds there)? However, 10 seconds would be too short in the sidecar mode, because we need to give enough time for the sidecar proxy to get configured by Pilot after BPF is configured for the endpoint, and Envoy's retry timeout is 20+ seconds.

I propose to wait for:

- 10 seconds in the node proxy mode; that should be sufficient for Envoy to get configured and ACK;
- 30 seconds in the sidecar mode; it's OK to wait longer and the total regeneration time can be >55 seconds, because endpoint creation is not blocked by regeneration completion in the sidecar mode.

File: pkg/envoy/xds/cache.go
Comment (rlenglet) [reply pkg/envoy/xds/cache.go:121]: This map will be passed as parameter `upsertedResources`, so I wanted to match that name.

File: pkg/envoy/xds/cache.go
Comment (rlenglet) [reply pkg/envoy/xds/cache.go:128]: Likewise, this slice will be passed as the `deletedNames` parameter, and I wanted to match that name.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
CI build failure (Cassandra runtime tests):
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6727/execution/node/92/log/
```
00:31:08.154 RuntimeCassandra 
00:31:08.154   Tests policy allowing all actions
00:31:08.154   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
00:31:41.422 STEP: Setting up policy: /home/vagrant/go/src/github.com/cilium/cilium/test/runtime/manifests/Policies-cassandra-allow-all.json
00:31:41.423 STEP: Inserting Value into Cassandra (no policy)
00:31:41.423 STEP: Reading values from Cassandra (no policy)
00:31:41.423 === Test Finished at 2018-09-26T00:08:03Z====
00:31:41.423 ===================== TEST FAILED =====================
00:31:41.423 cmd: sudo cilium endpoint list
00:31:41.423 Exitcode: 0 
00:31:41.423 Stdout:
00:31:41.423  	 ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])   IPv6                 IPv4            STATUS   
00:31:41.423 	            ENFORCEMENT        ENFORCEMENT                                                                                     
00:31:41.423 	 7175       Disabled           Disabled          4          reserved:health               f00d::a0f:0:0:1c07   10.15.123.251   ready   
00:31:41.423 	 23739      Enabled            Disabled          29706      container:id.cass-server      f00d::a0f:0:0:5cbb   10.15.122.18    ready   
00:31:41.423 	 58454      Disabled           Disabled          3814       container:id.cass-client      f00d::a0f:0:0:e456   10.15.24.209    ready   
00:31:41.423 	 
00:31:41.423 Stderr:
00:31:41.423  	 
00:31:41.423 
00:31:41.423 cmd: sudo cilium policy get
00:31:41.423 Exitcode: 0 
00:31:41.423 Stdout:
00:31:41.423  	 [
00:31:41.423 	   {
00:31:41.423 	     "endpointSelector": {
00:31:41.423 	       "matchLabels": {
00:31:41.423 	         "any:id.cass-server": ""
00:31:41.423 	       }
00:31:41.423 	     },
00:31:41.423 	     "ingress": [
00:31:41.423 	       {
00:31:41.423 	         "toPorts": [
00:31:41.423 	           {
00:31:41.423 	             "ports": [
00:31:41.423 	               {
00:31:41.423 	                 "port": "9042",
00:31:41.423 	                 "protocol": "TCP"
00:31:41.423 	               }
00:31:41.423 	             ],
00:31:41.423 	             "rules": {
00:31:41.423 	               "l7proto": "cassandra",
00:31:41.423 	               "l7": [
00:31:41.423 	                 {}
00:31:41.423 	               ]
00:31:41.423 	             }
00:31:41.423 	           }
00:31:41.423 	         ]
00:31:41.423 	       }
00:31:41.423 	     ]
00:31:41.423 	   }
00:31:41.423 	 ]
00:31:41.423 	 Revision: 15
00:31:41.423 	 
00:31:41.423 Stderr:
00:31:41.423  	 
00:31:41.423 
00:31:52.176 ===================== Existing AfterFailed =====================
00:31:52.176 <Checks>
00:31:52.176 Number of "context deadline exceeded" in logs: 0
00:31:52.176 ⚠️  Number of "level=error" in logs: 10
00:31:52.176 ⚠️  Number of "level=warning" in logs: 7
00:31:52.176 Number of "Cilium API handler panicked" in logs: 0
00:31:52.176 ⚠️  Number of "Goroutine took lock for more than" in logs: 12
00:31:52.176 
00:31:52.176 </Checks>
00:31:52.176 
00:31:52.177 [[ATTACHMENT|865c8e74_RuntimeCassandra_Tests_policy_allowing_all_actions.zip]]
00:31:52.177 
00:31:52.177 • Failure [43.968 seconds]
00:31:52.177 RuntimeCassandra
00:31:52.177 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
00:31:52.178   Tests policy allowing all actions [It]
00:31:52.178   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
00:31:52.178 
00:31:52.178   Unable to select from 'posts' table
00:31:52.178   Expected command: docker exec -i  cass-client cqlsh 10.15.122.18 --connect-timeout=200 -e "SELECT * FROM posts_db.posts;" 
00:31:52.178   To succeed, but it failed:
00:31:52.178   Exitcode: 1 
00:31:52.178   Stdout:
00:31:52.178    	 
00:31:52.178   Stderr:
00:31:52.178    	 Connection error: ('Unable to connect to any servers', {'10.15.122.18': ConnectionShutdown('Connection to 10.15.122.18 was closed',)})
00:31:52.178   	 
00:31:52.178   
00:31:52.178 
00:31:52.178   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/runtime/cassandra.go:144
```
```
STEP: Inserting Value into Cassandra (no policy)
time="2018-09-26T00:08:02Z" level=info msg="Executing command 'INSERT INTO posts_db.posts (username, creation, content) values ('alice', now(), 'Hello');'"
time="2018-09-26T00:08:02Z" level=debug msg="running command: docker exec -i  cass-client cqlsh 10.15.122.18 --connect-timeout=200 -e \"INSERT INTO posts_db.posts (username, creation, content) values ('alice', now(), 'Hello');\""
cmd: "docker exec -i  cass-client cqlsh 10.15.122.18 --connect-timeout=200 -e \"INSERT INTO posts_db.posts (username, creation, content) values ('alice', now(), 'Hello');\"" exitCode: 0 duration: 771.333215ms stdout:

STEP: Reading values from Cassandra (no policy)
time="2018-09-26T00:08:02Z" level=info msg="Executing command 'SELECT * FROM posts_db.posts;'"
time="2018-09-26T00:08:02Z" level=debug msg="running command: docker exec -i  cass-client cqlsh 10.15.122.18 --connect-timeout=200 -e \"SELECT * FROM posts_db.posts;\""
cmd: "docker exec -i  cass-client cqlsh 10.15.122.18 --connect-timeout=200 -e \"SELECT * FROM posts_db.posts;\"" exitCode: 1 duration: 627.913198ms stdout:

stderr:
Connection error: ('Unable to connect to any servers', {'10.15.122.18': ConnectionShutdown('Connection to 10.15.122.18 was closed',)})
```
This is an issue with the Cassandra parser:
```
Sep 26 00:08:02 runtime cilium-agent[14812]: level=debug msg="Queueing container event" containerID=82c0cbda72 event="exec_create: cqlsh 10.15.122.18 --connect-timeout=200 -e INSERT INTO posts_db.posts (username, creation, content) values ('alice', now(), 'Hello');" subsys=workload-watcher
Sep 26 00:08:02 runtime cilium-agent[14812]: level=debug msg="Queueing container event" containerID=82c0cbda72 event="exec_start: cqlsh 10.15.122.18 --connect-timeout=200 -e INSERT INTO posts_db.posts (username, creation, content) values ('alice', now(), 'Hello');" subsys=workload-watcher
Sep 26 00:08:02 runtime cilium-agent[14812]: level=debug msg="Queueing container event" containerID=82c0cbda72 event=exec_die subsys=workload-watcher
Sep 26 00:08:02 runtime cilium-agent[14812]: level=debug msg="Queueing container event" containerID=82c0cbda72 event="exec_create: cqlsh 10.15.122.18 --connect-timeout=200 -e SELECT * FROM posts_db.posts;" subsys=workload-watcher
Sep 26 00:08:03 runtime cilium-agent[14812]: level=debug msg="Queueing container event" containerID=82c0cbda72 event="exec_start: cqlsh 10.15.122.18 --connect-timeout=200 -e SELECT * FROM posts_db.posts;" subsys=workload-watcher
Sep 26 00:08:03 runtime cilium-agent[14812]: level=info msg="time=\"2018-09-26T00:08:03Z\" level=error msg=\"Compression flag set, unable to parse beyond the header\"" subsys=envoy-config threadID=23323
Sep 26 00:08:03 runtime cilium-agent[14812]: level=info msg="time=\"2018-09-26T00:08:03Z\" level=error msg=\"Request length of 1526726656 is greater than 256 MB\"" subsys=envoy-config threadID=23323
Sep 26 00:08:03 runtime cilium-agent[14812]: level=warning msg="[cilium_proxylib.cc:289] [C5] Cilium Network::OnIO: FILTER_POLICY_DROP Parser error" subsys=envoy-filter threadID=23346
```
Addressed in https://github.com/cilium/cilium/pull/5691.

**Comment by rlenglet** [comment]
test-me-please

**Comment by danwent** [comment]
Thanks for pointing out the runtime failure @rlenglet .   This is a legit bug that only surfaces when multiple reply requests as passed to OnData at the same time.  We handled the must requests scenario properly, but reply parsing was only added later in order to support prepared queries, and we do not yet have good runtime test coverage on prepared queries due to the fact that cqlsh cannot generate them.      There is a python cassandra SDK that can generate them, so I'll need to get some runtime tests based on that in the future, I suspect.

**Comment by rlenglet** [comment]
test-me-please

**Comment by jrajahalme** [comment]
There is substantial overlap now with this PR and #4535, would be good to discuss before merging either.

**Comment by rlenglet** [comment]
Successful CI build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6741/
Moving to `pending-review`, so **please review but please don't merge for now**. Merging this needs to be synchronized with #4535.

**Comment by rlenglet** [comment]
After discussion with @jrajahalme, removing commit https://github.com/cilium/cilium/pull/5451/commits/2b205fd30f25066aa1e0f3429f425b69b44fe45b ("Exclude BPF compilation from Envoy regeneration timeout") since it's not strictly required in this PR and it would conflict with https://github.com/cilium/cilium/pull/4535.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
CI build failure, due to monitor runtime test failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6752/execution/node/92/log/
```
00:16:13.746 RuntimeMonitorTest
00:16:13.747 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
00:16:13.747   With Sample Containers
00:16:13.747   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
00:16:13.747     multiple monitors [It]
00:16:13.748     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
00:16:13.748 
00:16:13.748     Expected
00:16:13.748         <string>: Listening for events on 2 CPUs with 64x4096 of shared memory
00:16:13.748         Press Ctrl-C to quit
00:16:13.748         <- endpoint 38870 flow 0x2eaea438 identity 1381->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 23707 flow 0x2eaea438 identity 1381->8223 state established ifindex lxcc3eaa16a861d: 10.15.17.19 -> 10.15.76.156 EchoRequest
00:16:13.748         <- endpoint 23707 flow 0x2eaea438 identity 8223->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 38870 flow 0x2eaea438 identity 8223->1381 state reply ifindex lxc479e2034445c: 10.15.76.156 -> 10.15.17.19 EchoReply
00:16:13.748         <- endpoint 38870 flow 0x2eaea438 identity 1381->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 23707 flow 0x2eaea438 identity 1381->8223 state established ifindex lxcc3eaa16a861d: 10.15.17.19 -> 10.15.76.156 EchoRequest
00:16:13.748         <- endpoint 23707 flow 0x2eaea438 identity 8223->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 38870 flow 0x2eaea438 identity 8223->1381 state reply ifindex lxc479e2034445c: 10.15.76.156 -> 10.15.17.19 EchoReply
00:16:13.748         <- endpoint 38870 flow 0x2eaea438 identity 1381->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 23707 flow 0x2eaea438 identity 1381->8223 state established ifindex lxcc3eaa16a861d: 10.15.17.19 -> 10.15.76.156 EchoRequest
00:16:13.748         <- endpoint 23707 flow 0x2eaea438 identity 8223->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 38870 flow 0x2eaea438 identity 8223->1381 state reply ifindex lxc479e2034445c: 10.15.76.156 -> 10.15.17.19 EchoReply
00:16:13.748         <- endpoint 38870 flow 0x2eaea438 identity 1381->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 23707 flow 0x2eaea438 identity 1381->8223 state established ifindex lxcc3eaa16a861d: 10.15.17.19 -> 10.15.76.156 EchoRequest
00:16:13.748         <- endpoint 23707 flow 0x2eaea438 identity 8223->0 state new ifindex 0: [unknown]
00:16:13.748         -> endpoint 38870 flow 0x2eaea438 identity 8223->1381 state reply ifindex lxc479e2034445c: 10.15.76.156 -> 10.15.17.19 EchoReply
00:16:13.748         
00:16:13.748     to contain substring
00:16:13.748         <string>: -> endpoint 38870 flow 0x2eaea438 identity 8223->1381 state reply ifindex lxc479e2034445c: [unknown]

00:16:13.748 
00:16:13.748     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/runtime/monitor.go:275
```

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Successful CI Build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6755/

File: pkg/envoy/xds/ack_test.go
Comment (aanm) [inline pkg/envoy/xds/ack_test.go:272]: Can you add a test that runs `upsert` 2 times and you want to revert the first upsert?

File: pkg/envoy/xds/cache.go
Comment (aanm) [inline pkg/envoy/xds/cache.go:134]: You can put this outside of this for loop and outside of the next for loop and have a if statement that does
```
if len(upsertedResources) != 0 || len(deletedNames) != 0{
cacheIsUpdated = true
}
```

File: pkg/endpoint/bpf.go
Comment (aanm) [inline pkg/endpoint/bpf.go]: Maybe these changes is in the wrong commit? I don't see a `e.Push`

Comment (ianvernon) [reply pkg/endpoint/bpf.go]: @aanm I think this is just the framework part of it - with just this commit, there aren't any functional changes since nothing is pushed yet.

File: pkg/endpoint/revert.go
Comment (ianvernon) [inline pkg/endpoint/revert.go]: unit tests ?

File: pkg/endpoint/revert.go
Comment (ianvernon) [inline pkg/endpoint/revert.go]: There's nothing wrong with this being in the endpoint package since endpoint is only consumer, but I could see this type being in its own separate package  (apart from the endpoint logger being used). It's pretty small though, so that's an argument in keeping it where it is now.

File: pkg/endpoint/revert.go
Comment (ianvernon) [inline pkg/endpoint/revert.go]: Unit tests?

File: pkg/proxy/proxy.go
Comment (ianvernon) [inline pkg/proxy/proxy.go]: if RevertFunc and FinalizeFunc were in their own packages, you wouldn't have to duplicate the types in proxy / endpoint packages.

File: pkg/proxy/proxy.go
Comment (ianvernon) [inline pkg/proxy/proxy.go]: Shouldn't we try to run all revert funcs if possible? And then return a list of errors? I guess we will have inconsistent state either way, though :/

**Review by ianvernon — COMMENTED** [review]
LGTM overall, just a few stylistic nitpicks. It would be nice to document what exactly the behavior of the datapath is once we hit these cases. I.e., if the reverts are successful, can we assume that the endpoint will have full connectivity as if nothing happened? I ask because of how proxy-ports are integrated into the PolicyMap. The main fix here is to not have Envoy itself be in a bad state, as opposed to having the datapath be OK for endpoints. Is that correct?

File: pkg/endpoint/bpf.go
Comment (jrajahalme) [inline pkg/endpoint/bpf.go]: Does it matter that `proxyID` could have been in the map previously and now it is just deleted?

**Review by jrajahalme — COMMENTED** [review]
Will review the xds pieces after rebase.

File: pkg/envoy/xds/ack_test.go
Comment (rlenglet) [reply pkg/envoy/xds/ack_test.go:272]: Done.

File: pkg/envoy/xds/cache.go
Comment (rlenglet) [reply pkg/envoy/xds/cache.go:134]: That's not the correct logic. See that if `!force`, `cacheIsUpdated` is set to `true` only when:
- A resource to update is inserted or different from the current version of the resource in the cache.
- A resource to delete actually already exists in the cache.

File: pkg/endpoint/bpf.go
Comment (rlenglet) [reply pkg/endpoint/bpf.go]: Correct, this is built in multiple commits.

File: pkg/proxy/proxy.go
Comment (rlenglet) [reply pkg/proxy/proxy.go]: Good idea, it's much cleaner. Done.

File: pkg/proxy/proxy.go
Comment (rlenglet) [reply pkg/proxy/proxy.go]: There are typically dependencies between the revert functions, so that's not possible.

File: pkg/endpoint/revert.go
Comment (rlenglet) [reply pkg/endpoint/revert.go]: Done.

File: pkg/endpoint/bpf.go
Comment (rlenglet) [reply pkg/endpoint/bpf.go]: Good catch. I updated to only delete the `proxyID` from `realizedRedirects` if it wasn't there before regeneration.

**Comment by rlenglet** [comment]
@jrajahalme I rebased and addressed review comments. Please do another review.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Successful CI build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6874/

**Review by ianvernon — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]
Went through the rebased code, still looking good.

**Comment by rlenglet** [comment]
I tested by hand that a NACK from Envoy results in the changes to proxy redirects and xDS network policies to be reverted.
This is ready to merge.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Successful CI build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6909/

**Comment by raybejjani** [comment]
@cilium/bpf Any chance someone could review this?

Alternatively, I can janitor merge this. I will likely time out in 24 hours and merge.

**Review by tgraf — APPROVED** [review]

**Pull request merged**
